File: src/structure/module/extractModuleName.ts
import path from "node:path";
import {Configuration} from "../../configuration/configuration.js";
import {removePrefix} from "../removePrefix.js";

function generateRelativeFileName(prefixes: string[], sourceFileName: string) {
    return removePrefix(sourceFileName, prefixes)
        .replace(/\.d\.ts$/, "")
        .replace(/\.ts$/, "")
}

export function extractModuleName(
    sourceFileName: string,
    configuration: Configuration,
) {
    const {inputRoots, libraryName} = configuration

    const relativeFileName = generateRelativeFileName(inputRoots, sourceFileName)

    return path.join(libraryName, relativeFileName)
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/module/extractModuleName.kt
package io.github.sgrishchenko.karakum.structure.module

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.structure.removePrefix
import js.array.ReadonlyArray
import node.path.path

private fun generateRelativeFileName(prefixes: ReadonlyArray<String>, sourceFileName: String): String {
    return removePrefix(sourceFileName, prefixes)
        .replace("\\.d\\.ts$".toRegex(), "")
        .replace("\\.ts$".toRegex(), "")
}

fun extractModuleName(
    sourceFileName: String,
    configuration: Configuration,
): String {
    val inputRoots = configuration.inputRoots
    val libraryName = configuration.libraryName

    val relativeFileName = generateRelativeFileName(inputRoots, sourceFileName)

    return path.posix.join(libraryName, relativeFileName)
}
